numba/core/typing/npydecl.py

```diff
@@ -1254,7 +1254,7 @@ class NumbaCArray(CallableTemplate):
     def generic(self):
         func_name = self.key.__name__
 
-        def typer(ptr, shape, dtype=types.none):
+        def typer(ptr, shape, dtype=None):
```

Comment:
This change needs some thinking about. I think the default now matches what an `@overload` would implement, but I'm not sure if this is correct WRT `@overload_glue`. There's also the question of whether the subsequent type check needs to be a `is_nonelike()` type check for the case of a `types.none` appearing, not sure if this is actually possible yet.

Reply:
Hm... that's a problem. It will be a `types.none` when called with `c_arr = numba.carray(arr, (3,), dtype=None)`.

Reply:
From OOB chat, time would be better spent translating this to `@overload` than trying to fix this edge case in `@overload_glue`, added this as issue: https://github.com/numba/numba/issues/8009